File: leetcode/tree/binary_tree_path.py
from typing import List
from common import TreeNode

class Solution:
    def binaryTreePaths(self, root: TreeNode) -> List[str]:
        if not root:
            return None
        res = []
        def helper(root, out):
            if not root.left and not root.right:
                res.append(out + str(root.val))
                return
            out += str(root.val) + "->"
            if root.left:
                helper(root.left, out)
            if root.right:
                helper(root.right, out)
        helper(root, "")
        return res

if __name__ == "__main__":
    root = TreeNode(3)
    root.left = TreeNode(9)
    root.right = TreeNode(20)
    root.right.left = TreeNode(15)
    root.right.right = TreeNode(7)
    print(Solution().binaryTreePaths(root))

File: leetcode/tree/binary_tree_path.go
package main

import (
	"fmt"
	"strconv"
)

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func binaryTreePaths(root *TreeNode) []string {
	var res []string
	helper(root, "", &res)
	return res
}

func helper(root *TreeNode, str string, out *[]string) {
	if root.Left == nil && root.Right == nil {
		str += strconv.Itoa(root.Val)
		*out = append(*out, str)
		return
	}
	str += strconv.Itoa(root.Val) + "->"
	if root.Left != nil {
		helper(root.Left, str, out)
	}
	if root.Right != nil {
		helper(root.Right, str, out)
	}
}

func main() {
	root := &TreeNode{3, nil, nil}
	root.Left = &TreeNode{9, nil, nil}
	root.Right = &TreeNode{20, nil, nil}
	root.Right.Left = &TreeNode{15, nil, nil}
	root.Right.Right = &TreeNode{7, nil, nil}
	out := binaryTreePaths(root)
	fmt.Println(out)
}
